### ISSUE TITLE
Enable manually computing SM covariance

### ISSUE THREAD

**Issue by DavidMStraub** [smelli/smelli#3]
If there is a problem with the SM covariance (e.g. when one adds observables manually), there is a warning 

> UserWarning: There was a problem loading the SM covariances. Please recompute them with `make_measurement`.

However the constructor of `GlobalLikelihood` actually then goes on to call `make_measurement` itself. While we could simply change the warning text, I think this is actually a problem because it prevents the user from making an informed choice about parameters such as `N`.

**Comment by DavidMStraub** [comment]
I realized that #2 made this a bit more complicated. We need to make sure the class can be instantiated if neither SM nor exp covariances exist.

**Comment by DavidMStraub** [comment]
A possible workaround is to add a `sm_cov_folder` argument analogous to `exp_cov_folder`. But I don't like it since we then have 4 arguements (the two folders , `N`, and `Nexp`) that clutter the `__init__` and don't have any effect normally.

**Comment by peterstangl** [comment]
> However the constructor of `GlobalLikelihood` actually then goes on to call `make_measurement` itself. While we could simply change the warning text, I think this is actually a problem because it prevents the user from making an informed choice about parameters such as `N`.

So do you think we should not call `make_measurement` in `__init__` in the case that the SM covariance has not been loaded?
This would probably mean that we would have to make sure that `make_measurement` is called by the user before any of the other methods can be called. 

One idea would be to set a variable `sm_cov_loaded = True` once the covariance has been successfully loaded. In this case, `make_measurement` can be called automatically in `__init__`. If, on the other hand, `sm_cov_loaded` is `False`, `make_measurement` will not be called automatically and  there should be an error raised if any of the methods of `GlobalLikelihood` is called, telling the user to first call `make_measurement`.
What do you think?

**Comment by DavidMStraub** [comment]
Yes that sounds good but I'm not sure how to generalize that to the exp covariance; here the problem is that by default, it is *not* loaded but computed (I insisted on this back then). So at the moment one *has* to supply `Nexp` to `__init__` if one wants to modify it.

In principle one could get rid of loading the exp covariance in `__init__` by enforcing the slightly more cumbersome

```python3
gl = GlobalLikelihood()
gl.load_exp_covariances(DIR)
gl.make_measurement()
```
rather than
```python3
gl = GlobalLikelihood(exp_cov_folder=DIR)
```
The downside of this is that the former version calls `make_measurement` twice and is thus slower, but not by much.

**Comment by peterstangl** [comment]
> Yes that sounds good but I'm not sure how to generalize that to the exp covariance; here the problem is that by default, it is _not_ loaded but computed (I insisted on this back then). So at the moment one _has_ to supply `Nexp` to `__init__` if one wants to modify it.

I don't see what the problem with the exp covariance is. If one wants to load the exp covariance, one provides the folder (and `Nexp` will be ignored). And if one does not want to load an exp covariance, but wants to use a higher precision for its computation, one provides `Nexp`.
One only has to make sure that `Nexp` is passed to `make_measurement` in the case when loading of the SM covariance fails.
Or am I missing something?

> In principle one could get rid of loading the exp covariance in `__init__` by enforcing the slightly more cumbersome
> 
> ```python
> gl = GlobalLikelihood()
> gl.load_exp_covariances(DIR)
> gl.make_measurement()
> ```
> 
> rather than
> 
> ```python
> gl = GlobalLikelihood(exp_cov_folder=DIR)
> ```
> 
> The downside of this is that the former version calls `make_measurement` twice and is thus slower, but not by much.

I do not understand what's the advantage of geting rid of loading the exp covariance in `__init__`. I think if one wants to load the exp covariance, it should not be computed first (like it is done in the above former version).

**Comment by DavidMStraub** [comment]
It's just a bit counterintuitive that the exp covariance is loaded by passing arguments to the constructor, while the SM covariance is loaded manually. But I realized only now that passing `Nexp` actually *does* have an effect in the default setting, so I think it is OK if we treat them differently.

**Comment by peterstangl** [comment]
Hm, yes, since one is loaded by default and the other one is computed by default, it is perhaps not possible to treat them completely equivalently.

Concerning the experimental covariance, one could think about the following use cases:
- **load default exp. cov.:** *N/A, there is no default exp. cov.*
- **load custom exp. cov.:** provide `exp_cov_folder` to `GlobalLikelihood`
- **compute exp. cov. with default precision:** default, nothing special to be done
- **compute exp. cov. with custom precision:** provide `Nexp` to `GlobalLikelihood`

Concerning the SM covariance, one could think about:
- **load default SM cov.:** default, nothing special to be done
- **load custom SM cov.:** first instantiate `GlobalLikelihood` (which includes loading the default SM cov.) and then call `load_sm_covariances` as well as `make_measurement`
- **compute SM cov. with default precision:** first instantiate `GlobalLikelihood` (which loads the default SM cov.) and then call `make_measurement` with `force=True`
- **compute SM cov. with custom precision:** first instantiate `GlobalLikelihood` (which includes loading the default SM cov.) and then call `make_measurement` with `force=True` and `N=...`

So the non-default cases concerning the SM covariance are always quite cumbersome and the default SM covariance will always be loaded first.
I wonder if we should make it easier to recompute and load custom theoretical covariances. Perhaps also BSM covariances (e.g. to check the assumptions made in fastfit).

**Comment by DavidMStraub** [comment]
Yes you are right. I guess that providing it all in `__init__` is the simplest solution after all.

What about the number of threads for the SM covariance? I think a `threads` argument would be misleading (most of `GlobalLikelihood` is not parallel). Add an argument with a different name such as `threads_sm_cov`?

**Comment by peterstangl** [comment]
> Yes you are right. I guess that providing it all in `__init__` is the simplest solution after all.

Perhaps yes. But I was also thinking about if it might make sense to only include a `sm_cov_folder` in the `__init__` (or `th_cov_folder` if one also wants to consider BSM cov. matrices) and use a completely separate class for actually calculating the covariances. A class `CovarianceMatrices` could be instantiated by `GlobalLikelihood` and could by default just load the sm cov. and compute the exp. cov.. Further, `GlobalLikelihood` could pass the arguments `sm_cov_folder`, `exp_cov_folder` and `Nexp` to this class if one wants to use custom matrices or increase the precision for the exp. cov. On the other hand, `CovarianceMatrices` could be used completely independently of `GlobalLikelihood` to compute any kind of covariance matrices (exp, SM, BSM) with any precision and number of threads etc. What do you think?

> What about the number of threads for the SM covariance? I think a `threads` argument would be misleading (most of `GlobalLikelihood` is not parallel). Add an argument with a different name such as `threads_sm_cov`?

If we include all the functionality in `GlobalLikelihood`, this would make sense I think.
Concerning the number of threads, I actually noticed that when computing the exp. cov., all my processor cores are used even though no `threads` argument was given. I also noticed that the computation is different for `FastFit` and `FastLikelihood` (using the new `MeasurementCovariance` class). How does the new computation of the exp. cov. in `MeasurementCovariance` actually use multi-threading?

**Comment by DavidMStraub** [comment]
Started to implement this but... what do we actually do if no SM cov is provided and loading the default one fails? (This happens e.g. when observables are added.) Throwing an error is not really an option since instantiation is necessary to call `make_measurement`. Or simply skip `make_measurement` in this case?

**Comment by DavidMStraub** [comment]
> A class `CovarianceMatrices` could be instantiated by `GlobalLikelihood` and could by default just load the sm cov. and compute the exp. cov.

But actually this will be almost the same use case as the `flavio.statistics.likelihood.MeasurementCovariance` and `flavio.statistics.likelihood.SMCovariance` are there for, so maybe one should not have too much duplication.

> how does the new computation of the exp. cov. in `MeasurementCovariance` actually use multi-threading?

Interesting, I suspect this is `np.linalg.inv` using a multithreaded BLAS, I haven't parallelized anything there.

**Comment by peterstangl** [comment]
> > A class `CovarianceMatrices` could be instantiated by `GlobalLikelihood` and could by default just load the sm cov. and compute the exp. cov.
> 
> But actually this will be almost the same use case as the `flavio.statistics.likelihood.MeasurementCovariance` and `flavio.statistics.likelihood.SMCovariance` are there for, so maybe one should not have too much duplication.

Yes, that's of course true. So should we then just add the optional arguments `sm_cov_folder`, `N`, `threads_sm_cov`? What about computing BSM cov. matrices?

> Started to implement this but... what do we actually do if no SM cov is provided and loading the default one fails? (This happens e.g. when observables are added.) Throwing an error is not really an option since instantiation is necessary to call `make_measurement`. Or simply skip `make_measurement` in this case?

If loading fails, I would issue a warning, skip `make_measurement` and set a variable `sm_cov_loaded=False`. All methods that use the covariances should check if `sm_cov_loaded==True` and throw an error otherwise.

**Comment by DavidMStraub** [comment]
> So should we then just add the optional arguments `sm_cov_folder`, `N`, `threads_sm_cov`

Yes I guess so. I already started and can open a PR soon.

> What about computing BSM cov. matrices?

Suprisingly, there seems to be no function `flavio.np_covariance`. This should be easily doable, but until then I think we should not bother here.

> If loading fails, I would issue a warning, skip `make_measurement` and set a variable `sm_cov_loaded=False`. All methods that use the covariances should check if `sm_cov_loaded==True` and throw an error otherwise.

Actually I think we don't have to do this since what will happen is that the first `FastLikelihood` instance will throw the `ValueError` "You need to call `make_measurement` first." which is exactly what the user should do. This way we don't have to add `try ... except` at several places in `GlobalLikelihood`.

**Comment by DavidMStraub** [comment]
> So should we then just add the optional arguments `sm_cov_folder`, `N`, `threads_sm_cov`

Actually I just realized that when we skip `make_measurement` after a failed load, the last two options are never needed.

**Comment by peterstangl** [comment]
> > So should we then just add the optional arguments `sm_cov_folder`, `N`, `threads_sm_cov`
> 
> Actually I just realized that when we skip `make_measurement` after a failed load, the last two options are never needed.

OK, but this would mean that if one wants to compute the SM cov, one first has to first instantiate `GlobalLikelihood` without special arguments, which tries loading the default SM cov. and maybe issues a warning and then one has to call `make_measurement` with `force=True`. But of course, otherwise we would also need an additional argument to tell `GlobalLikelihood` that it should recompute the SM cov. So i think it is actually fine like this

**Comment by peterstangl** [comment]
Solved by #4.
